src/libproc_macro/bridge.rs

```diff
@@ -0,0 +1,856 @@
+                    $oty(storage.$oty.alloc(x))
+                }
+            }
+
+            impl<S: server::Types> HandleToData<'_, S> for $oty {
+                type Data = S::$oty;
+                fn to_data(self, storage: &mut Storage<S>) -> Self::Data {
+                    let handle = self.0;
+                    mem::forget(self);
+                    storage.$oty.take(handle)
+                }
+            }
+
+            // FIXME(eddyb) not implementing `HandleToData` because that borrows mutably.
+            impl $oty {
+                #[allow(unused)]
+                fn to_data<S: server::Types>(&'s self, storage: &'s Storage<S>) -> &'s S::$oty {
+                    &storage.$oty[self.0]
+                }
+            }
+
+            impl<S: server::Types> HandleToData<'s, S> for &'s mut $oty {
+                type Data = &'s mut S::$oty;
+                fn to_data(self, storage: &'s mut Storage<S>) -> Self::Data {
+                    &mut storage.$oty[self.0]
+                }
+            }
+        )*
+
+        $(
+            #[repr(C)]
+            #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+            pub(crate) struct $ity(storage::Handle);
+            impl !Send for $ity {}
+            impl !Sync for $ity {}
+
+            impl<S: server::Types> HandleFromData<S> for $ity {
+                type Data = S::$ity;
+                fn from_data(x: Self::Data, storage: &mut Storage<S>) -> Self {
+                    $ity(storage.$ity.alloc(x))
+                }
+            }
+
+            impl<S: server::Types> HandleToData<'_, S> for $ity {
+                type Data = S::$ity;
+                fn to_data(self, storage: &mut Storage<S>) -> Self::Data {
+                    storage.$ity.take(self.0)
+                }
+            }
+        )*
+    }
+}
+define_handles! {
+    'owned:
+    TokenStream,
+    TokenStreamBuilder,
+    TokenStreamIter,
+    Group,
+    Literal,
+    SourceFile,
+    Diagnostic,
+
+    'interned:
+    Op,
+    Term,
+    Span,
+}
+
+// FIXME(eddyb) generate these impls by pattern-matching on the
+// names of methods - also could use the presence of `fn drop`
+// to distinguish between 'owned and 'interned, above.
+// Alternatively, special 'modes" could be listed of types in with_api
+// instead of pattern matching on methods, here and in server decl.
+
+impl Clone for TokenStream {
+    fn clone(&self) -> Self {
+        self.clone()
+    }
+}
+
+impl fmt::Display for TokenStream {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(&self.display())
+    }
+}
+
+impl Clone for TokenStreamIter {
+    fn clone(&self) -> Self {
+        self.clone()
+    }
+}
+
+impl Clone for Group {
+    fn clone(&self) -> Self {
+        self.clone()
+    }
+}
+
+impl Clone for Literal {
+    fn clone(&self) -> Self {
+        self.clone()
+    }
+}
+
+// FIXME(eddyb) `Literal` should not expose internal `Debug` impls.
```

Comment:
I've seen this FIXME in a few other places but it's probably not applicative here